src/main/java/com/direwolf20/buildinggadgets/common/building/CapabilityBlockProvider.java

```diff
@@ -0,0 +1,54 @@
+package com.direwolf20.buildinggadgets.common.building;
+
+import com.direwolf20.buildinggadgets.common.building.placement.SingleTypeProvider;
+import net.minecraft.init.Blocks;
+import net.minecraft.nbt.NBTBase;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTUtil;
+import net.minecraft.util.EnumFacing;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.capabilities.Capability.IStorage;
+import net.minecraftforge.common.capabilities.CapabilityInject;
+import net.minecraftforge.common.capabilities.CapabilityManager;
+
+import javax.annotation.Nonnull;
+
+public class CapabilityBlockProvider {
+
+    @CapabilityInject(IBlockProvider.class)
+    public static Capability<IBlockProvider> BLOCK_PROVIDER = null;
+
+    static IBlockProvider DEFAULT_AIR_PROVIDER = new SingleTypeProvider(Blocks.AIR.getDefaultState());
+
+    public CapabilityBlockProvider() {}
+
+    public static void register() {
+        CapabilityManager.INSTANCE.register(IBlockProvider.class, new IStorage<IBlockProvider>() {
+            @Nonnull
+            @Override
+            public NBTBase writeNBT(Capability<IBlockProvider> capability, IBlockProvider provider, EnumFacing facing) {
+                if (provider instanceof TranslationWrapper) {
+                    provider = ((TranslationWrapper) provider).getHandle();
```

Comment:
Do you have any suggestion of how could this be implemented? Personally I have no idea except adding a new interface called that specifies a wrapper

Reply:
Recursion

Reply:
If you'd not want you BlockProviders to be allow some kind of NBT serialisation, then you'd write a helper Method which does no more than if it is a translationwrapper, return the Method applied to the handle, if not return the input

Reply:
The nicer approach would be to allow for NBTSerialisation in the interface, as than the Wrapper could do it's job properly and it would not be necessary to use this getHandle Method.